extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DatasourceOptimizer.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.materializedview;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import io.druid.client.TimelineServerView;
+import io.druid.client.selector.ServerSelector;
+import io.druid.query.Query;
+import io.druid.query.TableDataSource;
+import io.druid.query.groupby.GroupByQuery;
+import io.druid.query.spec.MultipleIntervalSegmentSpec;
+import io.druid.query.timeseries.TimeseriesQuery;
+import io.druid.query.topn.TopNQuery;
+import io.druid.timeline.TimelineObjectHolder;
+import org.joda.time.Interval;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Consumer;
+
+public class DatasourceOptimizer 
+{
+  private static final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private static ConcurrentHashMap<Derivative, AtomicLong> derivativesHitCount = new ConcurrentHashMap<>();
```

Comment:
`DataSourceOptimizer` is a singleton, and  I use `static` because `optimize` method is a static method.
>> Why don't we simply making a singleton instance of this?

Do you mean I should write another class (e.g. DataSourceOptimizerMetrics) to do record these states.

Reply:
Oh, you're right. It's singleton. Then, I wonder why you made the `optimize()` method static. Usually static methods are useful when a class doesn't have to keep any states (like util classes). But, `DataSourceOptimizer` does keep states (that is, metrics).